### PR TITLE
Allow pasting using Cmd+V in wxTextCtrl with wxTE_PASSWORD style

### PR DESCRIPTION
**Pull request by vadz**
For some reason known only to Apple, NSSecureTextField allows pasting text
into it using its standard context menu, but not using the standard Cmd+V
keyboard shortcut. Moreover, apparently the control does something special to
disable it because pressing Cmd+V does result in a call to
-[NSSecureTextField control:textView:doCommandBySelector:], but with a dummy
"noop:" selector.

Detect this specific situation and check if we're called while handling Cmd+V
event and, if this is indeed the case, do paste the text into the control.

While it could be argued that this changes the platform behaviour, it seems
very hard, if not impossible, to imagine a situation in which this would be a
problem while not being able to easily paste into password fields is
definitely a real usability bug.

---

@vslavik I'd appreciate any comments about this, TIA!

### CODE REVIEW

**Comment by vslavik** [comment]
It’s a gross hack (but you know that), but much more importantly, it doesn’t seem to be needed.

It sounded too stupid to be true even by Apple’s standard and a quick search confirmed that [pasting into `NSSecureTextField` is supposed to work](http://stackoverflow.com/a/5036946/237188) (that article says 10.5+, I’m using 10.12). I added the control into a native application to test and indeed, I could paste with Cmd+V without a problem.

So next I thought this must be some bad interaction with wx’s way of doing things (wouldn’t be the first time). But no, it works in Poedit too. Both with a wx standard `wxID_PASTE` menu item present and with patched-in native menu item that invokes the standard `paste:` selector (unsurprisingly as both are the same thing).

The last thing I tried was to remove standard Edit menu items completely and then yes, `NSTextField` pasting still worked, but `NSSecureTextField` one didn’t. But I wasn't even sure pasting with Cmd+V *is* supposed to work in this case in the first place — *all* native applications do have the Edit menu with its standard items.

…so I tried that in a native application too, removed the Paste menu item and guess what: pasting with Cmd+V did not work with either `NSSecureTextField` or `NSTextField`.

In conclusion, I think the actual bug in wx is that `wxTextCtrl` must be doing some shenanigans to handle Cmd+V itself instead of letting the standard menu do it, or its commands handling is interfering somehow. That’s probably what causes the weird `noop:` selector too.

**Comment by vadz** [comment]
The original problem is that `Cmd+V` doesn't work in a password field in a dialog (not `wxPasswordEntryDialog`, but something similar) while it does work in a normal text entry field in the same dialog. Everything else aside, I think we can agree that this is wrong and `Cmd+V` should be made to work in this situation somehow. This is _not_ the native behaviour because pasting password doesn't work in the Keychain standard dialog neither -- but IMO this is wrong and we don't lose anything by deviating from the native behaviour here.

I'm really not sure how else am I supposed to make it work though. I don't see any code dealing with it in `wxTextCtrl` itself and when I put a breakpoint at `-[wxNSTextField controlTextDidChange]`, it seems to be called by some native code, i.e. if there is any wx code in play here, it doesn't appear in the stack any more. So what could I do for `wxNSSecureTextField`? I guess I could intercept the original event and call `Paste()` from it, but I'm not actually sure it's going to work (`NSSecureTextField` has many weird restrictions and doesn't seem to behave in the same way as the standard `NSTextField` at all which, BTW, results in several other bugs that can be seen in the "text" sample, but this is a topic for another day...) and what would be the advantage of it compared to this solution, as long as it works?

To summarize, this localized hack results in the behaviour we want without, AFAICS, any drawbacks. And I don't see any alternative way of achieving this behaviour. Do you see a better approach?

Finally, and this is just speculation, but I think that `NSSecureTextField` does translate `Cmd+V` into `noop:` itself internally and that this is just a leftover from the pre-OS X 10.5 version that forbade pasting into it at all.

**Comment by vslavik** [comment]
> Everything else aside, I think we can agree that this is wrong and Cmd+V should be made to work in this situation somehow.

I agree the behavior should be consistent between the two. I lack your conviction that adding *more* non-native behavior is the answer. I don’t know why Cmd+V works in `wxTextCtrl` in absence of the Edit➞Paste menu entry, but it is *not* native behavior.

> This is *not* the native behaviour because pasting password doesn't work in the Keychain standard dialog neither

I see you dismissed my previous comment, the reference and the lengths I went to to check the native behaviour :-( I can only recommend that you actually do try it — as I did — create a native project in Xcode, add password field in interface builder and try it out. You’ll discover I wasn’t lying to you.

Re Keychain access: it’s the Keychain window that isn’t behaving as `NSSecureTextField` normally does, and understandably so: it’s limited to entry of your login password (which unlike other passwords, you *really* shouldn’t be storing anywhere) and it’s out of process and running with no menu of its own. (So it too actually *is* behaving consistently with “normal” use: no Paste menu, no pasting.)

> To summarize, this localized hack results in the behaviour we want without, AFAICS, any drawbacks. 

I agree that the change seems harmless and there’s little chance of it breaking something — that is, *if* this code doesn’t get called in the presence of a proper `wxID_PASTE` menu item (does it?). But there’s at least one drawback if you don’t use `wxID_PASTE`: the Edit menu doesn’t flash when you use the shortcut, which is an expected visual cue when you use a shortcut on Mac. And if you add the Paste item to fix that, you don’t need the hack anymore...

> I'm really not sure how else am I supposed to make it work though.

Sorry for not spelling it out explicitly, I thought it was sufficiently implied by what I wrote: make your application code well-behaved on Mac w.r.t. text editing, i.e. add an Edit menu with Cut/Copy/Paste items. That’s how it works in all native applications, users expect to have it (and so apparently do the native controls) and it fixes the problem.

**Comment by vadz** [comment]
Point by point:

1. Native behaviour: sorry, I was wrong then, I thought it was a built-in behaviour for the text fields to handle `Cmd+V`. If it isn't, it's really too bad for Mac users but who am I to judge...
1. I didn't dismiss it at all, thanks for checking it. All I said was that I had also checked it in the native context and saw that this behaviour wasn't specific to wx. I hadn't realized that Keychain behaved differently (and, to be honest, I don't completely buy your explanation: Keychain can be used for passwords other than the login one too!). Do you know of any other native password prompts? E.g. does pasting work in Safari when you log it to some site?
1. Flashing "Edit" menu: good point, I didn't think about it and didn't check it, will do.
1. Most important: no, sorry, I hadn't realized that your proposed solution was to just add the "Edit" menu. I suppose it will work (but I'd need to test it unless you had done it already in wx context?), but I don't really like it very much: my current application is not document-oriented at all, it's basically a single dialog and it doesn't have any menus (of course, under Mac it gets the standard menu bar with the standard items, but this happens automatically). I definitely don't want to have any menus under Windows, they would be completely useless there, so I would need to either use conditional compilation which is unpalatable or change wxOSX to add the "Edit" menu automatically. The latter might be a good idea -- or not, I'm really not sure. It definitely seems like a very big hammer for this particular nail though... Do you really think this should be done, and is safe to do? If yes, we could discuss this further, but if no, then I still don't have any other solution for my case.

To summarize,  assuming the patch doesn't interfere with "Edit|Paste" (I'll test it, but I am relatively confident it won't), it still seems like a useful change to me. The only alternative is to automatically create the standard "Edit" menu for all wx applications, and this could be useful too, if it's so important, but there are several potential problems with doing it...

**Comment by vslavik** [comment]
> If it isn't, it's really too bad for Mac users but who am I to judge…

It’s normal for *all* Mac apps to have the global menu and to have the Edit submenu it in, with Paste item. Therefore, Cmd+V **does work** in all native apps as you would expect. 

> I don't completely buy your explanation: Keychain can be used for passwords other than the login one too!)

Keychain is used to *store* other passwords. But the only password you *enter* in Keychain’s UI is the one to unlock the keychain, i.e. the one password you need to know, not have stored… 

> Do you know of any other native password prompts?

Again, just create a trivial application in Xcode. But yes, you can see it in stock apps too, e.g. Mail’s account password entry or System Preferences➞iCloud➞Account Details.

> I suppose it will work (but I'd need to test it unless you had done it already in wx context?), 

I did — see my 1st comment.

> but I don't really like it very much

You don’t like Macs very much in the first place 🙂 … but this is how native applications behave, like it or not. Even those that don’t need it have the Edit menu with these disabled (extremely rare, because show me an app that has no text entry *anywhere*… remember menus are global).

> or change wxOSX to add the "Edit" menu automatically. The latter might be a good idea -- or not, I'm really not sure.

In indeed might, [there’s a ton needed in there](https://github.com/vslavik/poedit/blob/672fce0eecb593022cfb06e53a53acd5a8a45088/src/edapp.cpp#L1182), but it may also be more trouble than it’s worth. Especially when as I suspect most wx users aren’t as picky about nativeness as I am.

> If yes, we could discuss this further, but if no, then I still don't have any other solution for my case.

Well, you do dismiss the simplest and right (from user POV) one as “unpalatable”, so there’s that…

But sure, if this doesn’t break anything, I don’t see the harm as accommodation for less-than-native wx apps (after all, Cmd+C/X/V/A are handled too already).

**Comment by vadz** [comment]
> Well, you do dismiss the simplest and right (from user POV) one as “unpalatable”, so there’s that…

I'm not sure if I expressed myself correctly, but in my case, when there is no menu whatsoever in the MSW version, this "simple and right" solution would involve having `#ifdef`s in the code. Surely we're not going to say that this is really acceptable?

> But sure, if this doesn’t break anything

I've retested with an "Edit" menu in the text sample and it seems to work correctly, i.e. `Cmd+V` still works and does flash the menu, so I'll merge this and will return to the idea of adding the standard "Edit" menu by default if/when we find something else affected by its absence.

Thanks!

**Comment by vslavik** [comment]
> Surely we're not going to say that this is really acceptable?

Oh, not only do I find it very much acceptable, I’m going to say it’s *necessary and required* to litter your wx code with `#ifdefs` for wxOSX. But we have very different priorities, I go out of my way to pretend to be a native application.

**Comment by vadz** [comment]
I think that any `#ifdef PLATFORM` check in the code is a failure of wxWidgets. Granted, in practice they're unavoidable but surely the best solution would avoid them.

In this case it does seem to mean adding the "Edit" menu automatically, but, as I wrote, there are several potential issues with doing it...

**Comment by vadz** [comment]
Merged as fe685bc4c6dff358366f6c5cbbe46b003232c9ec
